File: packages/provide-uterm-server/src/provide/uterm/server/bridge/hub/router_redaction.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Frame-field redaction helpers for the message router.

Pure, stateless helpers extracted from
:mod:`provide.uterm.server.bridge.hub.router_impl`. They take a
:class:`StreamRedactor` and return redacted copies of wire frames /
nested payloads; they never mutate their inputs and have no dependency
on :class:`MessageRouter` or the hub. ``router_impl`` re-exports
:func:`_redact_value` and :func:`_redact_frame_fields` so the public
``router_impl`` namespace (and the tests that import these from it) is
unchanged.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from provide.uterm.server.bridge.hub.redaction import StreamRedactor


# Defensive recursion bound for _redact_value. Wire frames are JSON-decoded
# (acyclic) and shallow in practice; this cap only guards against a
# pathologically deep structure causing a RecursionError. A value below the cap
# is returned as-is rather than redacted — secrets that deep don't occur in real
# snapshot/analysis payloads, and failing closed on depth (returning raw) is the
# conservative choice vs. raising on a hot broadcast path.
_REDACT_MAX_DEPTH = 32


def _redact_value(value: Any, redactor: StreamRedactor, _depth: int = 0) -> Any:
    """Recursively redact string values inside nested dict/list structures.

    Strings are redacted; dicts and lists are walked (values/elements only,
    not dict keys); all other scalars (int/float/bool/None) are returned
    unchanged. Recursion is capped at ``_REDACT_MAX_DEPTH`` — values deeper than
    that are returned verbatim. Input is never mutated; new containers are built.
    """
    if isinstance(value, str):
        return redactor.redact(value)
    if _depth >= _REDACT_MAX_DEPTH:
        return value
    if isinstance(value, dict):
        return {k: _redact_value(v, redactor, _depth + 1) for k, v in value.items()}
    if isinstance(value, list):
        return [_redact_value(v, redactor, _depth + 1) for v in value]
    return value


def _redact_frame_fields(msg: dict[str, Any], redactor: StreamRedactor) -> dict[str, Any]:
    """Return a copy of *msg* with its terminal-content string fields redacted.

    Applies role-scoped redaction to the content-bearing fields of term,
    snapshot, and analysis frames; other frame types are returned unchanged.
    """
    mtype = msg.get("type")
    if mtype == "term":
        return {**msg, "data": redactor.redact(str(msg.get("data", "")))}
    if mtype == "snapshot":
        out = dict(msg)
        out["screen"] = redactor.redact(str(msg.get("screen", "")))
        raw_tail = msg.get("raw_tail")
        if isinstance(raw_tail, str):
            out["raw_tail"] = redactor.redact(raw_tail)
        # prompt_detected is a dict that can carry the matched prompt text
        # (which may include secrets); redact its nested string values. None /
        # absent stays as-is (_redact_value returns scalars unchanged).
        if "prompt_detected" in out:
            out["prompt_detected"] = _redact_value(out["prompt_detected"], redactor)
        return out
    if mtype == "analysis":
        out = dict(msg)
        out["formatted"] = redactor.redact(str(msg.get("formatted", "")))
        raw = msg.get("raw")
        if isinstance(raw, str):
            out["raw"] = redactor.redact(raw)
        elif isinstance(raw, (dict, list)):
            # A structured raw was previously shipped verbatim; redact nested
            # string values so secrets inside dict/list raw don't egress.
            out["raw"] = _redact_value(raw, redactor)
        return out
    return msg


__all__ = ["_REDACT_MAX_DEPTH", "_redact_frame_fields", "_redact_value"]

File: packages/provide-uterm-go/hub/router_redaction.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

// Frame-field redaction helpers for the message router. Port of
// provide.uterm.server.bridge.hub.router_redaction. These are pure, stateless
// helpers: they take a [StreamRedactor] and return redacted COPIES of wire
// frames / nested payloads; they never mutate their inputs.

// redactMaxDepth is the defensive recursion bound for redactValue. Wire frames
// are JSON-decoded (acyclic) and shallow in practice; a value below the cap is
// returned as-is rather than redacted — failing closed on depth (returning raw)
// is the conservative choice on a hot broadcast path. Port of _REDACT_MAX_DEPTH.
const redactMaxDepth = 32

// RedactFrameFields is the concrete [Redactor] that wires the real
// [StreamRedactor] into the hub's output-policy seam. It builds a redactor from
// rules and returns a role-redacted copy of msg. Assign it to
// TermHubConfig.Redactor to activate output redaction when a gate yields rules.
//
// This mirrors the Python broadcast path, which calls
// _redact_frame_fields(msg, StreamRedactor(rules)) directly (Python has no
// pluggable-redactor seam — StreamRedactor is hardcoded there).
func RedactFrameFields(msg map[string]any, rules []RedactionRule) map[string]any {
	return redactFrameFields(msg, NewStreamRedactor(rules))
}

// redactValue recursively redacts string values inside nested map/slice
// structures. Strings are redacted; maps and slices are walked (values/elements
// only, not map keys); all other scalars are returned unchanged. Recursion is
// capped at redactMaxDepth. The input is never mutated. Port of _redact_value.
func redactValue(value any, r *StreamRedactor, depth int) any {
	if s, ok := value.(string); ok {
		return r.Redact(s)
	}
	if depth >= redactMaxDepth {
		return value
	}
	switch v := value.(type) {
	case map[string]any:
		out := make(map[string]any, len(v))
		for k, vv := range v {
			out[k] = redactValue(vv, r, depth+1)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, vv := range v {
			out[i] = redactValue(vv, r, depth+1)
		}
		return out
	default:
		return value
	}
}

// redactFrameFields returns a copy of msg with its terminal-content string fields
// redacted, for term / snapshot / analysis frames; other frame types are
// returned unchanged (same reference). Port of _redact_frame_fields.
func redactFrameFields(msg map[string]any, r *StreamRedactor) map[string]any {
	switch str(msg["type"]) {
	case "term":
		out := copyMap(msg)
		out["data"] = r.Redact(str(msg["data"]))
		return out
	case "snapshot":
		out := copyMap(msg)
		out["screen"] = r.Redact(str(msg["screen"]))
		if rawTail, ok := msg["raw_tail"].(string); ok {
			out["raw_tail"] = r.Redact(rawTail)
		}
		// prompt_detected can carry the matched prompt text (which may include
		// secrets); redact its nested string values. Absent/scalar stays as-is.
		if pd, ok := out["prompt_detected"]; ok {
			out["prompt_detected"] = redactValue(pd, r, 0)
		}
		return out
	case "analysis":
		out := copyMap(msg)
		out["formatted"] = r.Redact(str(msg["formatted"]))
		switch raw := msg["raw"].(type) {
		case string:
			out["raw"] = r.Redact(raw)
		case map[string]any:
			out["raw"] = redactValue(raw, r, 0)
		case []any:
			out["raw"] = redactValue(raw, r, 0)
		}
		return out
	default:
		return msg
	}
}

// copyMap returns a shallow copy of m (a new top-level map sharing the original
// values). Redaction helpers build new containers so the stored frame is never
// mutated.
func copyMap(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}
